### PR TITLE
Task 5 (Integration with S3)

### PR DESCRIPTION
**Pull request by furiousme**
**Important!** I've started work on the next task, that's why if something doesn't work as expected, chances are I just messed around :) Please contact me and I'll deploy the relevant app version. Thanks!

[Task](https://github.com/rolling-scopes-school/aws/blob/main/aws-developer/05_integration_with_s3/task.md)

[importProductsFile lambda](https://kkd22giog0.execute-api.eu-west-1.amazonaws.com/prod/import)

[FE PR](https://github.com/furiousme/nodejs-aws-shop-react/pull/4)

**Total: 100**

## Basic scope (70 points)

- [x]  AWS CDK Stack contains configuration for `importProductsFile` function
- [x]  The `importProductsFile` lambda function returns a correct response which can be used to upload a file into the S3 bucket
- [x]  Frontend application is integrated with `importProductsFile` lambda
- [x]  The `importFileParser` lambda function is implemented and AWS CDK Stack contains configuration for the lambda

## Additional (optional) tasks (30 points)

- [x]  **+10** **(All languages)** - `importProductsFile` lambda is covered by _unit tests_.
- [x]   You should consider to mock S3 and other AWS SDK methods so not trigger actual AWS services while unit testing.
- [x]  **+10** **(All languages)** - `importFileParser` lambda is covered by _unit tests_.
- [x]  **+10** **(All languages)** - At the end of the stream the lambda function should move the file from the `uploaded` folder into the `parsed` folder (`move the file` means that file should be copied into a new folder in the same bucket called `parsed`, and then deleted from `uploaded` folder)

_**Note:**_  I decided not to use the same bucket for parsed files just to be safe as this is not a best practice. I'm not focused much on points, so please evaluate this requirement as you see fit. 

Screenshot 2024-06-30 at 22 12 08
Screenshot 2024-06-30 at 22 17 52
Screenshot 2024-06-30 at 22 18 38

### CODE REVIEW

File: ImportService/lib/handlers/import-file-parser/import-file-parser.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
+const csv = require("csv-parser");
+
+import { FIXME } from './../../../../ProductService/models/index';
+
+import { CopyObjectCommand, DeleteObjectCommand, GetObjectCommand, S3Client } from "@aws-sdk/client-s3";
+import { Product } from 'aws-cdk-lib/aws-servicecatalog';
```
Comment (hazardsoft) [inline ImportService/lib/handlers/import-file-parser/import-file-parser.ts]: Seems to be unrelevant to the project, please refer to docs to read about `Product` in AWS ecosystem (https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_servicecatalog-readme.html)

File: ImportService/lib/import_service-stack.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+      autoDeleteObjects: true,
+      cors: [{
+        allowedOrigins: ["*"],
+        allowedHeaders: ["*"],
+        maxAge: 3600,
+        allowedMethods: [HttpMethods.PUT, HttpMethods.POST]
+      }]
+    });
+
+    const destinationBucket = new Bucket(this, 'ProductsFilesDestinationBucket', {
```
Comment (hazardsoft) [inline ImportService/lib/import_service-stack.ts:36]: `uploaded`/`parsed` prefixes should be in the same bucket accordingly to the reqs.

File: ImportService/lib/import_service-stack.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        proxy: false,
+      defaultCorsPreflightOptions: {
+        allowMethods: [
+          CorsHttpMethod.GET,
+          CorsHttpMethod.POST,
+          CorsHttpMethod.PUT,
+          CorsHttpMethod.PATCH,
+          CorsHttpMethod.DELETE,
+          CorsHttpMethod.OPTIONS,
+        ],
```
Comment (hazardsoft) [inline ImportService/lib/import_service-stack.ts]:
```suggestion
        allowMethods: [
          CorsHttpMethod.GET,
        ],
```

**Review by hazardsoft — APPROVED** [review]

**Review by hazardsoft — APPROVED** [review]

File: ImportService/lib/handlers/import-file-parser/import-file-parser.ts (hunk 1 above)
Comment (furiousme) [reply ImportService/lib/handlers/import-file-parser/import-file-parser.ts]: hmm auto import...

File: ImportService/lib/import_service-stack.ts (hunk 2 above)
Comment (furiousme) [reply ImportService/lib/import_service-stack.ts:36]: Yep, I know. This is why I left this note in the description)

> Note: I decided not to use the same bucket for parsed files just to be safe as this is not a best practice. I'm not focused much on points, so please evaluate this requirement as you see fit.

File: ImportService/lib/import_service-stack.ts (hunk 3 above)
Comment (furiousme) [reply ImportService/lib/import_service-stack.ts]: Absolutely redundant, agreed ))

Thanks!

**Pull request merged**